[Daniel Sim] iP

danielsimre opened the pull request:
# Duke
> “Life is too complicated not to be orderly.” – Martha Stewart [(source)](https://dansilvestre.com/productivity-quotes)

Whenever your workload starts to pile up, have you ever felt:
- confused?
- disorganized?
- hopeless when you finally realize how much work you have?

Worry no further! Duke is a tool that aims to help you organize your work better, thereby increasing your overall efficiency when tackling your tasks.

### Why should you use Duke?
1. It's **easy to use**
2. It's **lightweight** (text-based)
3. ~~And most importantly,~~ it's **free**

### Features include (but are not limited to):
- [x] Manage tasks, deadlines and events
- [x] Save/load functionality
- [x] Get all deadlines/events happening on a certain date
- [x] Get all deadlines/events occurring within x hours/days/months from now
- [ ] Get all overdue tasks (Planned feature)

### Below is some sample code from Duke:
```java
public static void main(String[] args) {
    Path filePath = Paths.get(System.getProperty("user.dir"), "data", "tasks.txt");
    new Duke(filePath).run();
}
```

So _what_ are you waiting for? Download Duke right now! :grin:

Commit:
Add Gradle support

danielsimre committed:
Change welcome message

danielsimre committed:
Add basic echo and exit functionality

danielsimre committed:
Change scanner to take whole line as input

danielsimre committed:
Add add task and list tasks functionality

danielsimre committed:
Add Task.java and Task class functionality

danielsimre committed:
Change chatbot to use Task objects instead of String

danielsimre committed:
Refactor Task.java functions

danielsimre committed:
Add done functionality to chatbot

danielsimre committed:
Add Todo, Deadline, Event classes and functionality

danielsimre committed:
Add functionality for chatbot to add todo/deadline/event tasks

danielsimre committed:
Modify input.txt and EXPECTED.txt to enable automated testing

danielsimre committed:
Add DukeException.java to represent exceptions specific to Duke

danielsimre committed:
Add error handling and usage of DukeException

danielsimre committed:
Update text ui test to account for error handling

danielsimre committed:
Extract code in main into smaller functions

danielsimre committed:
Add delete functionality to chatbot

danielsimre committed:
Update automated testing to test delete functionality

danielsimre committed:
Refactor input commands into an enum called Command

danielsimre committed:
Fix indentation for switch case statements

danielsimre committed:
Update .gitignore to ignore savefile

danielsimre committed:
Update .gitignore to ignore /data/ directory

danielsimre committed:
Add save functionality to chatbot

danielsimre committed:
Add loading from savefile functionality on chatbot startup

danielsimre committed:
Add date and time attributes to deadline tasks

danielsimre committed:
Add date and time attributes to event

danielsimre committed:
Add get task by date functionality to chatbot

danielsimre committed:
Add ability to get upcoming due tasks

danielsimre committed:
Merge branch 'branch-Level-7' into master
Add save/load functionality to master branch.

danielsimre committed:
Merge branch 'branch-Level-8' into master

Add date/time related functionality to Duke.

danielsimre committed:
Throw error if user inputs 1 time for event

danielsimre committed:
Refactor according to coding standard

danielsimre committed:
Abstract out common code into new classes

danielsimre committed:
Update runtest.bat to use new commands

danielsimre committed:
Sort common files into packages

danielsimre committed:
Update runtest.bat to work with packages

danielsimre committed:
Add unit test for Parser.java

danielsimre committed:
Ignore META-INF folder

danielsimre committed:
Add javadocs to all non-command classes

danielsimre committed:
Standardise import statement order

danielsimre committed:
Standardise method order in classes

danielsimre committed:
Add find by keyword functionality to chatbot

danielsimre committed:
Merge branch 'branch-A-JavaDoc' into master
Add javadocs to master branch.

danielsimre committed:
Merge branch 'branch-A-CodingStandard' into master

Standardise code on master branch.

danielsimre committed:
Merge branch 'branch-Level-9' into master

Add find functionality to master branch.

danielsimre committed:
Update javadocs for non-command classes/methods

danielsimre committed:
Add javadocs for Command classes

danielsimre committed:
Remove unneccessary imports

Harry-Gao-H commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Object to represent the user's task list (e.g. add/delete/mark as done)
+     */
+    private TaskList tasks;
```
`tasks` here is an object of TaskList class instead of a collection of tasks. So perhaps it is better to name it `taskList`?

darren2pro replied:
This is a valid point. But I also think that "tasks" is appropriate too because it shows a plural form already.

danielsimre replied:
Thank you both for the comments! I agree with Harry that taskList is an more appropriate name, as the variable refers to a singular TaskList object, so I will make the changes accordingly.

Harry-Gao-H commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,272 @@
+            // Add Deadline task
+            String errorMessage = "Command must be in the format: [taskName] /by "
+                    + "[date(YYYY-MM-DD)] [time(HH:MM)].";
+            String[] splitInput = Parser.splitWith(input, 9, " /by ", errorMessage);
```
Should `splitInput` be in plural form?

darren2pro replied:
I think this is a valid point.

danielsimre replied:
Yup, thank you for catching this. The variable has been renamed to splitInputs.

Harry-Gao-H commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,272 @@
+            String[] splitInput = Parser.splitWith(input, 9, " /by ", errorMessage);
+            String taskName = splitInput[0];
+            errorMessage = "Date and time must be in the format: YYYY-MM-DD HH:MM.";
+            String[] dateTime = Parser.splitWith(splitInput[1], 0, " ", errorMessage);
```
Should `dateTime` be in plural form?

darren2pro replied:
I think this is a valid point too.

danielsimre replied:
Thank you for catching this. The variable has been renamed to dateTimeInputs.

darren2pro reviewed (commented):
Your coding standards look good! 👍

danielsimre committed:
Fix coding standard violations

danielsimre committed:
Merge branch 'add-gradle-support' into master
Add gradle support to Duke.

danielsimre committed:
Set up gradle in project

danielsimre committed:
Setup CheckStyle in project

danielsimre committed:
Fix checkstyle related errors

danielsimre committed:
Add dependencies for JavaFX

danielsimre committed:
Add JavaFX GUI to chatbot

danielsimre committed:
Add FXML to project

danielsimre committed:
Change label in dialog to text

danielsimre committed:
Remove unneeded parts in Ui class

danielsimre committed:
Connect Duke to JavaFx Gui

danielsimre committed:
Merge branch 'branch-level-10' into master
Add JavaFX UI to Duke.

danielsimre committed:
Add varargs to addTask function

danielsimre committed:
Merge branch 'branch-A-Varargs' into master
Add varargs to addTask function in TaskList.

danielsimre committed:
Add assertions to Duke

Duke currently does not use any assertions. (Only exception handling)

Let's,
* Edit the build.gradle file to enable assertions
* Add an assertion to getResponse in Duke.java to ensure that Duke
never returns a blank response
* Add an assertion to addTask in TaskList.java to check that no
non-adding commands will call that method

Assertions help to catch mistakes made in the code rather than by the
user, so this will help to improve the overall quality of the program.

danielsimre committed:
Modify code to comply with Code Quality standards

The code amongst all classes violates many code quality standards,
such as use of magic literals, lack of SLAP and long functions.

By fixing these code quality errors, it will be easier for others
(and myself) to quickly and better understand what the code is doing
at any given point.

Let's
* Replace all magic literals with constants with descriptive names.
* Abstract out code in functions to satisfy SLAP and reduce length of
  functions.
* Add meaningful comments to explain the why/what of code, and remove
comments that explain the how.
* Add guard clauses to certain functions to make the happy path clearer.

This will hopefully greatly reduce the amount of code quality violations
and make it much easier to understand and read.

danielsimre committed:
Merge pull request #1 from danielsimre/branch-A-CodeQuality

Modify code to comply with Code Quality standards

danielsimre committed:
Merge branch 'master' into branch-A-Assertions

Merge new changes related to code quality.

danielsimre committed:
Merge pull request #2 from danielsimre/branch-A-Assertions

Add assertions to Duke

danielsimre committed:
Improve find command to be case insensitive

Find command is case sensitive.

If the user enters a key phrase, the find command only returns
tasks that fit the phrase exactly. (e.g. find book does not return
the task Return Book)

Let's change containsPhrase method in Task to be case insensitive.

This will solve the aforementioned problem of case sensitive keyphrases.

Note: This commit is made on this branch so that I can create a pull
request for branch-A-Streams, since my code already uses streams
and hence was identical to the master branch.

danielsimre committed:
Merge pull request #3 from danielsimre/branch-A-Streams

Improve find command to be case insensitive

danielsimre committed:
Remove text ui tests

danielsimre committed:
Set up Continuous Integration

danielsimre committed:
Fix folder structure of github actions

danielsimre committed:
Add ability to specify command shortcuts

Users can only use the commands specified in the program.

Adding the ability for users to specify their own shortcuts
(or aliases) for commands will improve the user's workflow by
making the syntax more friendly overall.

Let's
* Add a class to handle aliases
* Add methods to save and load aliases to and from a config file
* Modify certain methods in TaskList to take into consideration
aliases of variable length

Having a dedicated class to handle aliases will ensure that
other classes would not get too long, as well as serving as
an encapsulation of alias-related functionality so that any
future modifications made to how aliases are handled will
not cause a cascade of modifications in other classes.

danielsimre committed:
Merge pull request #4 from danielsimre/branch-C-FriendlierSyntax

Add ability to specify command shortcuts

danielsimre committed:
Add Result class to project

Currently, the execute method for all commands returns a TaskList
object.

This is a sign of inadequate abstraction, as it does not make much
logical sense for all commands to return a task list, especially for
alias-related commands that do not modify the task list. In addition,
AliasHandler and TaskList currently contain a message attribute to
store the output message, which is a potential sign of repeated code.

Let's,
* create a new class, Result, that stores an AliasHandler object and
TaskList object, as well as an output message string.
* make all commands return Result instead of TaskList.
* remove the message attribute from TaskList and AliasHandler.

This would provide another layer of abstraction, as well as making
the code easier to understand. (The return value of executing a
Command is a Result.) It also reduces code duplication, as
there is now only one message attribute. (In Result)

danielsimre committed:
Remove unused TaskList constructor

danielsimre committed:
Improve GUI of app

danielsimre committed:
Change chatbot name to Energy

danielsimre committed:
Fix regex bug in due command

danielsimre committed:
Update README.md for repo

danielsimre committed:
Fix scrollbar not appearing

danielsimre committed:
Add Ui.png to docs folder

danielsimre committed:
Update user guide

danielsimre committed:
Fix picture in user guide

danielsimre committed:
Fix formatting errors in user guide

danielsimre committed:
Improve code quality

danielsimre committed:
Fix image loading error
